src/connection/stream.rs

```diff
@@ -414,5 +396,46 @@ impl Shared {
 
         current // Return the previous stream state for informational purposes.
     }
+
+    /// Calculate the number of additional window bytes the receiving side
+    /// should grant the sending side via a window update message.
+    ///
+    /// Returns `None` if too small to justify a window update message.
+    ///
+    /// Note: Once a caller successfully sent a window update message, the
+    /// locally tracked window size needs to be updated manually by the caller.
+    pub(crate) fn next_window_update(&mut self) -> Option<u32> {
+        if !self.state.can_read() {
+            return None;
+        }
+
+        let new_credit = match self.config.window_update_mode {
+            WindowUpdateMode::OnReceive => {
+                debug_assert!(self.config.receive_window >= self.window);
+                let bytes_received = self.config.receive_window.saturating_sub(self.window);
+                bytes_received
+            },
+            WindowUpdateMode::OnRead => {
+                debug_assert!(self.config.receive_window >= self.window);
+                let bytes_received = self.config.receive_window.saturating_sub(self.window);
+                let buffer_len: u32 = self.buffer.len()
```

Comment:
As far as I know `self.buffer.len()` is always a `fold()` on all chunks. Doing that essentially on every invocation of `poll_read` or `poll_next` makes me a bit uncomfortable (i.e. more frequently than it is currently done, I think once for each received data frame). Since each chunk comes from a data frame, these may be relatively small and we may even want to have smaller data frames rather than making them as large as possibly the window size, I wonder if the `Chunks` buffer should not better explicitly track its total size as chunks are `push`ed and `pop`ped.

Reply:
> Doing that essentially on every invocation of poll_read or poll_next makes me a bit uncomfortable

I profiled a 10 second benchmark run, using 10 streams with 10_000 messages of 256 byte each on an unconstrained network. I would imagine this scenario to cause many calls to `Chunks::len`. The amount of `Chunk`s in `Chunks` alternates between 0 and ~400. Below is the [flamegraph](https://github.com/flamegraph-rs/flamegraph) svg (as a .zip as Github doesn't allow svg). Indeed the call to `fold` shows up, taking around 1% of the overall CPU time.


[flamegraph.zip](https://github.com/paritytech/yamux/files/5958408/flamegraph.zip)

Running with the same settings as above, but instead using the gbit-lan scenario the `fold` calls take less than 0.2% of the overall CPU time.

> I wonder if the Chunks buffer should not better explicitly track its total size as chunks are pushed and popped.

One could explicitly track the size of all chunks through `push` and `pop`. Problem is that `Chunks` allows mutable access to the front chunk through `front_mut` for example used [in `Stream as AsyncRead`](https://github.com/paritytech/yamux/blob/90720d796ef66b64f6453d8342c8f786ea5611ef/src/connection/stream.rs#L273-L285). One could keep the length for all chunks and subtract the offset (bytes already read) of the first chunk on each call to `Chunks::len`.

Running the same benchmark with the diff below causes the `fold` call in `Chunks::len` to no longer show up at all. I am not sure this optimization is worth the additional complexity. In case we would like to optimize this code path, I would expect a switch to a single ring-buffer per `Stream` to be more fruitful.

What do you think @romanb?

```diff
diff --git a/src/chunks.rs b/src/chunks.rs
index c22308e..2f97424 100644
--- a/src/chunks.rs
+++ b/src/chunks.rs
@@ -17,24 +17,24 @@ use std::{collections::VecDeque, io};
 /// [`Chunk`] elements.
 #[derive(Debug)]
 pub(crate) struct Chunks {
-    seq: VecDeque<Chunk>
+    seq: VecDeque<Chunk>,
+    len: usize,
 }
 
 impl Chunks {
     /// A new empty chunk list.
     pub(crate) fn new() -> Self {
-        Chunks { seq: VecDeque::new() }
+        Chunks { seq: VecDeque::new(), len: 0 }
     }
 
     /// The total length of bytes contained in all `Chunk`s.
     pub(crate) fn len(&self) -> Option<usize> {
-        self.seq.iter().fold(Some(0), |total, x| {
-            total.and_then(|n| n.checked_add(x.len()))
-        })
+        Some(self.len - self.seq.front().map(|c| c.offset()).unwrap_or(0))
     }
 
     /// Add another chunk of bytes to the end.
     pub(crate) fn push(&mut self, x: Vec<u8>) {
+        self.len += x.len();
         if !x.is_empty() {
             self.seq.push_back(Chunk { cursor: io::Cursor::new(x) })
         }
@@ -42,7 +42,9 @@ impl Chunks {
 
     /// Remove and return the first chunk.
     pub(crate) fn pop(&mut self) -> Option<Chunk> {
-        self.seq.pop_front()
+        let chunk = self.seq.pop_front();
+        self.len -= chunk.as_ref().map(|c| c.len() + c.offset()).unwrap_or(0);
+        chunk
     }
```

Reply:
I did not consider the `front_mut` use, though the patch you've shown for explicit tracking of the length looks simple enough to me, i.e. I'd probably do it, but if you consider the impact shown by your measurements to be negligible, feel free to leave it for now. Thanks for the follow-up.

Reply:
> though the patch you've shown for explicit tracking of the length looks simple enough to me

Sounds good then. Done in https://github.com/paritytech/yamux/pull/109/commits/d4398f0fa23870e66a3e2a69fb99092a3cfb1e77.